spark/src/main/scala/org/apache/spark/sql/delta/hooks/AutoCompact.scala

```diff
@@ -0,0 +1,257 @@
+          tableId,
+          autoCompactRequest.allowedPartitions
+        )
+        metrics
+      } catch {
+        case e: Throwable =>
+          logError("Auto Compaction failed with: " + e.getMessage)
+          throw e
+      } finally {
+        if (AutoCompactUtils.reservePartitionEnabled(spark)) {
+          AutoCompactPartitionReserve.releasePartitions(
+            tableId,
+            autoCompactRequest.allowedPartitions
+          )
+        }
+      }
+    } else {
+      Seq.empty[OptimizeMetrics]
+    }
+  }
+
+
+  /**
+   * Launch Auto Compaction jobs if there is sufficient capacity.
+   * @param spark The spark session of the parent transaction that triggers this Auto Compaction.
+   * @param deltaLog The delta log of the parent transaction.
+   * @return the optimize metrics of this compaction job.
+   */
+  private[delta] def compact(
+      spark: SparkSession,
+      deltaLog: DeltaLog,
+      catalogTable: Option[CatalogTable],
+      partitionPredicates: Seq[Expression] = Nil,
+      opType: String = OP_TYPE,
+      maxDeletedRowsRatio: Option[Double] = None)
+  : Seq[OptimizeMetrics] = recordDeltaOperation(deltaLog, opType) {
+    recordDeltaOperation(deltaLog, s"$opType.execute") {
+      val txn = deltaLog.startTransaction(catalogTable)
+      val optimizeContext = DeltaOptimizeContext(
+        isPurge = false,
+        minFileSize = spark.conf.get(DeltaSQLConf.DELTA_AUTO_COMPACT_MIN_FILE_SIZE),
+        maxFileSize = spark.conf.get(DeltaSQLConf.DELTA_AUTO_COMPACT_MAX_FILE_SIZE),
+        maxDeletedRowsRatio = maxDeletedRowsRatio
+      )
+      val rows = new OptimizeExecutor(spark, txn, partitionPredicates, Seq(), true, optimizeContext)
+        .optimize()
+      val metrics = rows.map(_.getAs[OptimizeMetrics](1))
+      recordDeltaEvent(deltaLog, s"$opType.execute.metrics", data = metrics.head)
+      metrics
+    }
+  }
+
+}
+
+/**
+ * Post commit hook for Auto Compaction.
+ */
+case object AutoCompact extends AutoCompactBase
+/**
+ * A trait describing the type of Auto Compaction.
+ */
+sealed trait AutoCompactType {
+  val configValueStrings: Seq[String]
+  /**
+   * This is the default maximum file size for Auto Compaction.
+   * 1. MAX_FILE_SIZE is configurable and defaults to 16 MB unless overridden.
+   * 2. MIN_FILE_SIZE is configurable and defaults to MAX_FILE_SIZE / 2 unless overridden.
+   * Note: User can use DELTA_AUTO_COMPACT_MAX_FILE_SIZE to override this value.
+   */
+  val defaultMaxFileSize: Long = 16 * 1024 * 1024
+}
+
+object AutoCompactType {
+
+  private[hooks] val DISABLED = "false"
+
+  /**
+   * This represents the legacy auto-compact behaviour:
+   * 1. MAX_FILE_SIZE is configurable and defaults to 128 MB unless overridden.
+   * 2. MIN_FILE_SIZE is configurable and defaults to MAX_FILE_SIZE / 2 unless overridden.
+   * Note: User can use DELTA_AUTO_COMPACT_MAX_FILE_SIZE to override this value.
+   */
+  case object Enabled extends AutoCompactType {
```

Comment:
Should we have two distinct objects? Enabled and Legacy to make it clear?

Reply:
Legacy is a, well, legacy name, there are only `Enabled`, or `Disabled`